src/Microsoft.Azure.SignalR.Management/ServiceManagerBuilder.cs

```diff
@@ -84,16 +99,56 @@ public ServiceManagerBuilder WithCallingAssembly()
             return this;
         }
 
+        internal ServiceManagerBuilder ConfigureServices(Action<IServiceCollection> configureAction)
+        {
+            _configureAction = configureAction;
+            return this;
+        }
+
         /// <summary>
         /// Builds <see cref="IServiceManager"/> instances.
         /// </summary>
         /// <returns>The instance of the <see cref="IServiceManager"/>.</returns>
         public IServiceManager Build()
         {
-            return _services.AddSignalRServiceManager()
-                .AddSingleton(_services.ToList() as IReadOnlyCollection<ServiceDescriptor>)
-                .BuildServiceProvider()
+            var serviceCollection = new ServiceCollection().Add(_services);
+            _configureAction?.Invoke(serviceCollection);
+            serviceCollection.AddSingleton(_services.ToList() as IReadOnlyCollection<ServiceDescriptor>);
+            return serviceCollection.BuildServiceProvider()
                 .GetRequiredService<IServiceManager>();
         }
+
+        /// <summary>
+        /// Builds <see cref="ServiceHubContext"/> instances.
+        /// </summary>
+        /// <returns>The instance of the <see cref="IServiceManager"/>.</returns>
+        internal async Task<ServiceHubContext> CreateHubContextAsync(string hubName, CancellationToken cancellationToken)
+        {
+            //add requried services
+            using var serviceProvider = _services.BuildServiceProvider();
+            var transportType = serviceProvider.GetRequiredService<IOptions<ServiceManagerOptions>>().Value.ServiceTransportType;
+            var services = new ServiceCollection().Add(_services);
+            services.AddHub(hubName, transportType);
+            _configureAction?.Invoke(services);
+            services.AddSingleton(services.ToList() as IReadOnlyCollection<ServiceDescriptor>);
+
+            //build
+            var serviceHubContext = services.BuildServiceProvider()
+                .GetRequiredService<ServiceHubContextImpl>();
+
+            //initialize
+            var connectionContainer = serviceHubContext.ServiceProvider.GetService<IServiceConnectionContainer>();
+            if (connectionContainer != null)
+            {
+                await connectionContainer.ConnectionInitializedTask.OrTimeout(cancellationToken);
+            }
+
+            if (cancellationToken.IsCancellationRequested)
+            {
+                await serviceHubContext?.DisposeAsync();
```

Comment:
ConnectionInitializedTask can throw, wrap this into finally?

Reply:
I think we should dispose only if exception thrown, so wrap them into catch.

Reply:
do you want to also throw when `cancellationToken.IsCancellationRequested`